flagsmith/models.go

```diff
@@ -237,3 +237,147 @@ func MakeFeatureResourceDataFromClientFeature(clientFeature *flagsmithapi.Featur
 	}
 	return resourceData
 }
+
+type Condition struct {
+	Operator types.String `tfsdk:"operator"`
+	Property types.String `tfsdk:"property"`
+	Value    types.String `tfsdk:"value"`
+}
+func (c *Condition) ToClientCondition() *flagsmithapi.Condition {
+	return &flagsmithapi.Condition{
+		Operator: c.Operator.Value,
+		Property: c.Property.Value,
+		Value:    c.Value.Value,
+	}
+}
+
+func MakeConditionFromClientCondition(clientCondition * flagsmithapi.Condition) Condition {
+	return Condition{
+		Operator: types.String{Value: clientCondition.Operator},
+		Property: types.String{Value: clientCondition.Property},
+		Value:    types.String{Value: clientCondition.Value},
+	}
+
+}
+type NestedRule struct {
+	Conditions []Condition `tfsdk:"conditions"`
+	Type      types.String `tfsdk:"type"`
+}
+func (t *NestedRule) ToClientRule() *flagsmithapi.Rule {
+	conditions := make([]flagsmithapi.Condition, 0)
+	for _, condition := range t.Conditions {
+		conditions = append(conditions, *condition.ToClientCondition())
+	}
+	return &flagsmithapi.Rule{
+		Conditions: conditions,
+		Type:       t.Type.Value,
+	}
+}
+func MakeNestedRuleFromClientRule(clientRule * flagsmithapi.Rule) NestedRule {
+	var conditions []Condition
+	for _, clientCondition := range clientRule.Conditions {
+		conditions = append(conditions, MakeConditionFromClientCondition(&clientCondition))
+	}
+	return NestedRule{
+		Conditions: conditions,
+		Type:       types.String{Value: clientRule.Type},
+	}
+
+}
+
+type Rule struct {
+	Type types.String `tfsdk:"type"`
+	Rules []NestedRule `tfsdk:"rules"`
+	Conditions []Condition `tfsdk:"conditions"`
+}
+
+func (r *Rule) ToClientRule() *flagsmithapi.Rule {
+	var conditions []flagsmithapi.Condition
+	for _, condition := range r.Conditions {
+		conditions = append(conditions, *condition.ToClientCondition())
+	}
+	var rules []flagsmithapi.Rule
+	for _, rule := range r.Rules {
+		rules = append(rules, *rule.ToClientRule())
+	}
+	return &flagsmithapi.Rule{
+		Type: r.Type.Value,
+		Rules: rules,
+		Conditions: conditions,
+	}
+}
+
+func MakeRuleFromClientRule(clientRule * flagsmithapi.Rule) Rule {
+	rule := Rule{
+		Type: types.String{Value: clientRule.Type},
+	}
+	for _, clientCondition := range clientRule.Conditions {
+		rule.Conditions = append(rule.Conditions, MakeConditionFromClientCondition(&clientCondition))
+	}
+	for _, clientSubRule := range clientRule.Rules {
+		rule.Rules = append(rule.Rules, MakeNestedRuleFromClientRule(&clientSubRule))
+	}
+	return rule
+}
+
+type SegmentResourceData struct {
```

Comment:
Why `SegmentResourceData` and not just `Segment` ? Seems like we're doing this above for e.g. `FeatureResourceData` but I'm not entirely sure what the difference is here versus, e.g. `Condition` or `Rule`?

Reply:
The convention that I am following is to add `ResourceData` to the model that represents the resource and use the type name without any suffix/prefix for subtypes/nested types(Condition/Rule, FeatureStateValue etc..) .

Reply:
Ok, makes sense.